pkg/network/tracer/connection/tracer.go

```diff
@@ -96,716 +74,11 @@ type Tracer interface {
-	telemetry.NewCounter(connTracerModuleName, "iteration_aborts", []string{}, "Counter measuring how many times ebpf iteration of connection map was aborted"),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-	atomic.NewInt64(0),
-}
-
-type tracer struct {
-	m *manager.Manager
-
-	conns          *maps.GenericMap[netebpf.ConnTuple, netebpf.ConnStats]
-	tcpStats       *maps.GenericMap[netebpf.ConnTuple, netebpf.TCPStats]
-	tcpRetransmits *maps.GenericMap[netebpf.ConnTuple, uint32]
-	config         *config.Config
-
-	// tcp_close events
-	closeConsumer *tcpCloseConsumer
-	// tcp failure events
-	failedConnConsumer *failure.TCPFailedConnConsumer
-
-	removeTuple *netebpf.ConnTuple
-
-	closeTracer func()
-	stopOnce    sync.Once
-
-	ebpfTracerType TracerType
-
-	exitTelemetry chan struct{}
-
-	ch *cookieHasher
-}
-
-// NewTracer creates a new tracer
-func NewTracer(config *config.Config, _ telemetryComponent.Component) (Tracer, error) {
-	mgrOptions := manager.Options{
-		// Extend RLIMIT_MEMLOCK (8) size
-		// On some systems, the default for RLIMIT_MEMLOCK may be as low as 64 bytes.
-		// This will result in an EPERM (Operation not permitted) error, when trying to create an eBPF map
-		// using bpf(2) with BPF_MAP_CREATE.
-		//
-		// We are setting the limit to infinity until we have a better handle on the true requirements.
-		RLimit: &unix.Rlimit{
-			Cur: math.MaxUint64,
-			Max: math.MaxUint64,
-		},
-		MapSpecEditors: map[string]manager.MapSpecEditor{
-			probes.ConnMap:                           {MaxEntries: config.MaxTrackedConnections, EditorFlag: manager.EditMaxEntries},
-			probes.TCPStatsMap:                       {MaxEntries: config.MaxTrackedConnections, EditorFlag: manager.EditMaxEntries},
-			probes.TCPRetransmitsMap:                 {MaxEntries: config.MaxTrackedConnections, EditorFlag: manager.EditMaxEntries},
-			probes.PortBindingsMap:                   {MaxEntries: config.MaxTrackedConnections, EditorFlag: manager.EditMaxEntries},
-			probes.UDPPortBindingsMap:                {MaxEntries: config.MaxTrackedConnections, EditorFlag: manager.EditMaxEntries},
-			probes.ConnectionProtocolMap:             {MaxEntries: config.MaxTrackedConnections, EditorFlag: manager.EditMaxEntries},
-			probes.ConnectionTupleToSocketSKBConnMap: {MaxEntries: config.MaxTrackedConnections, EditorFlag: manager.EditMaxEntries},
-		},
-		ConstantEditors: []manager.ConstantEditor{
-			boolConst("tcpv6_enabled", config.CollectTCPv6Conns),
-			boolConst("udpv6_enabled", config.CollectUDPv6Conns),
-		},
-		DefaultKProbeMaxActive: maxActive,
-		BypassEnabled:          config.BypassEnabled,
+// NewTracer returns a new Tracer
+func NewTracer(cfg *config.Config, telemetryComp telemetry.Component) (Tracer, error) {
+	if cfg.EnableEbpfless {
+		return newEbpfLessTracer(cfg)
```

Comment:
Could the ebpf-less tracer be in a separate package and implement the `Tracer` interface? That was the original intent of that interface: to allow different implementations. Then you wouldn't have the `git blame` issue Lee was mentioning.

Creating `pkg/network/tracer/ebpfless` for the ebpf-less tracer.

Reply:
I did try that, and I could not really sort out the import cycles without making this PR bigger than it already is. Perhaps in a followup I can attempt that again.

The `git blame` issue is not as bad as we initially thought. I can see the blame showing up in github across the renames.